### ISSUE TITLE
Optimizing data for Display Names

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#3260]
The DisplayNames component comes with a large amount of data. It is the largest locale-specific data in ICU and will also likely be the largest in ICU4X.

There are a few things that make DisplayNames interesting:

1. The majority of the display names are probably not useful to carry for most clients. For example, users speaking Japanese are more likely to need the translation for the Katakana script than the translation for the Cherokee script. We should explore something like japanext and likelysubtagsext where we have a core set and an extended set.
2. Regional variants often override only a small number of strings. For example, en-GB and en-US might be equivalent for all region names except for one or two. This doesn't play nicely with the deduplication mechanism we've thusfar relied on.

CC @snktd @robertbastian @markusicu

**Comment by sffc** [comment]
Discuss with:

- @sffc 
- @Manishearth 
- @younies

**Comment by sffc** [comment]
Discussed on 2023-07-04. We will use the auxiliary key model, similar to currency formatter (#1441), which resolves the issues in the OP.

**Comment by sffc** [comment]
Options:

1. Fine grained data marker attributes
2. Core/Extended

Discussion:

- @zbraniecki Seems like it would be a footgun if clients aren't loading display names that they need.
- @sffc How would you draw the line?
- @zbraniecki Not really sure.
- @zbraniecki Most clients are not in a position to decide whether "Tovalu" should be included. It seems option 2 is easier for clients, if we can make a reasonable set. We are in a better position to draw this line than our customers.
- @sffc Option 1 is better for data deduplication. It's possible though that we could do something similar to how we reduced data size for time zones (where we allow fallback to root).
- @sffc My main worry with Option 1 is its impact on compile times with so many finely sliced data marker attributes.

**Comment by sffc** [comment]
Trying to organize my thoughts here.

There are two different ways we are discussing about reducing data size:

1. Deduplicating identical strings
    - Example: en-GB and en-US and fr have the same translation for Panama. We'd like to ship that translation only once.
2. Removing cold strings
    - Example: the display name of Uganda in Korean maybe doesn't need to be in a minimal data set.

These two paths can be explored mostly independently, though some solutions impact both.

## Deduplicating identical strings

Two general paths for this which are compatible with the ICU4X architecture:

1. Make every string its own data marker attribute. This is what we decided to do last summer, https://github.com/unicode-org/icu4x/issues/3260#issuecomment-1621296519/, and it is what we currently do for the experimental unit and currency formats.
    - Pro: Smallest postcard data size
    - Pro: Allows for cold strings to be removed via datagen (solves problem 2)
    - Pro: Easy to implement
    - Con: Finely sliced data markers increase build times and baked data size (https://github.com/unicode-org/icu4x/issues/5230)
    - Con: The data provider needs to be queried for every individual display name (as opposed to one load that returns a map)
2. Load multiple payloads, such as "locale plus root" or "locale plus script", which have been deduplicated. This is what @robertbastian suggested in https://github.com/unicode-org/icu4x/issues/3260#issuecomment-1497146025 and similar to what he implemented to deduplicate time zone names in https://github.com/unicode-org/icu4x/pull/5751
    - Pro: Doesn't test the boundaries of the Rust compiler
    - Pro: Should be fairly efficient for display name lookup
    - Con: Leaves some data size on the table (unclear how much)
    - Con: A bit harder to implement
    - Con: Increases stack size of DisplayNames type
    - Con: May favor Latin-script languages, but see https://github.com/unicode-org/icu4x/issues/5901

Is there another approach I'm missing?

## Removing cold strings

Two ways to go about this:

1. Utilize data marker attributes (see above)
    - All pros and cons as above
    - Pro: The set of strings included for each locale is fully customizable by the client, though we could still ship default configurations
    - Con: A non-default set of strings requires running datagen
2. Introduce Minimal/Core/Extended keys, similar to LocaleExpander.
    - Pro: Clients do not need to fiddle with datagen in order to meet their requirements
    - Pro: Puts the i18n team in charge of deciding what constitutes a "cold" string
    - Con: The choices offered to clients are coarse (such as minimal, core, or extended) and not customizable
    - Con: Increases stack size of DisplayNames type
    - Con: May reduce performance relative to not having the separate keys

Thoughts on any of the above?

@Manishearth @hsivonen

**Comment by sffc** [comment]
> While I am in favor of allowing clients to fully customize which strings they get, I think removing cold strings is prone to being confusing for people using our default data?

I'm not too worried about confusion with our default data. I think our default data should always be at least the "core" data set, which I'm considering to be the cartesian product of all `modern` locales. If we take the multiple-keys approach, the constructors can have names such as `new_minimal`, `new`, and `new_extended`.

> I'd rather attempt to deduplicate.

Yes, deduplication is something we should do regardless of how/if we decide to handle cold strings.

> Multiple payloads sounds like the okay-but-suboptimal solution, and we can plan for that whilst I figure out how fundamental the rust compiler limits are.

Ideally we work out the compiler fundamentals before implementing a solution. It seems unlikely we'll have this ready in 2024, but it would be nice to have a path forward and implement this in 2025.

> > Con: The data provider needs to be queried for every individual display name (as opposed to one load that returns a map)
> 
> I actually think this can be a pro! I think this depends on the usage patterns, and I could see us even having different types of display names formatters geared towards different use cases. The use case of "show the user the country they are in" is different from "show the users a dropdown of countries".
> 
> We're going to have a bit of recurring tension between "optimize data for multiple use cases at once" vs "make different data models for different use cases, but then people with both use cases have to avoid accidentally duplicating data". We've had this question crop up before for BidiAuxiliaryProperties.

True. What I had in mind when I wrote my comment was more like, "if you need one name at a time, a map lookup is about as fast as a data provider lookup, but if you need an iterator over all names, a map lookup is probably faster."

**Comment by Manishearth** [comment]
Yep, 100%.

**Comment by sffc** [comment]
> I don't like the cold-data thing. A user with a Korean UI and address in Germany is just as important as a user with a German UI and an address in Germany, even if there are more of the latter.

It's about tradeoffs. @zbraniecki likes to use the following thought experiment: in an environment with a fixed amount of space, what produces the better outcome for global users: including these cold display name strings, or including a whole new locale?

I do think we should keep the cold strings in the default configuration. But, I also think we are in a position to give this lever to clients who need it.

> We also don't have any data on which combinations are "hot".

CLDR has _some_ data on usage patterns between languages and regions. The data could probably use some improvement.

**Comment by hsivonen** [comment]
> the display name of Uganda in Korean maybe doesn't need to be in a minimal data set.

I think it's bad to slice up the data like this. When a South Korean app asks a user for their country-or-region of residence as part of user registration (let's assume for the sake of simplicity here that it's legitimate to ask for such a thing), why should the menu exclude Uganda?

AFAICT, splicing up the DisplayNames data like this doesn't make much sense, since DisplayNames makes the most sense to use when you need to deal with a large set of possible names and especially if translating them is outside the core competence of the app's UI localizers. If the app only needs to deal with a small set of translatable strings that are familiar to the translators, the app doesn't need DisplayNames.

Continuing with the Korean example: Anecdotally, if a South Korean app has non-Korean UI languages, the UI languages are Korean, English, Japanese, and Chinese (without clearly saying Simplified or Traditional). That's a small set, but it doesn't follow that it makes sense to provide DisplayNames pruned to a small set that's somehow geographically proximate plus English. The app does not need DisplayNames—hot or cold—to get labels for an in-app UI for switching between these. It can use whatever mechanism if uses for its UI strings in general.

(And making English have a broader set of supported names than Korean on the assumption that people whose country-or-region of residence is in the cold set for Korean would use the English UI anyway would fail cases like Koreans who have moved abroad and are visiting registering for an account with the UI in Korean but country-or-region of residence in the cold set. So that would be bad.)

Having an API that works for presumed-hot values but not for presumed-cold values makes for an unreliable API. If developers perceive an API as too unreliable to use, then even the utility per size of even the hot set becomes worse.

As far as the JavaScript ECMA-402 side goes, as I said in the TG2 meeting, I think it would be better to not ship DisplayNames than to ship DisplayNames in an unreliable state.

**Comment by Manishearth** [comment]
- @sffc Two ways to reduce data size:
    1. Deduplicate identical strings
    2. Remove cold strings
- @sffc I want to focus on (1). I had two approaches:
    1. Use finely-sliced data markers
    2. Load multiple payloads
- @sffc Ideally we could use finely-sliced data markers, but they have lot of problems of their own, as discussed at length in https://github.com/unicode-org/icu4x/issues/5230
- @robertbastian Finely-sliced data markers might not be a good API. We might not know the regions to load until runtime.
- @sffc We previously discussed a design that a DateTime formatter which only supports a single time zone, but the design we chose allows all time zones for formatting. In the case of display names, it's okay to say that you have the data marker when you choose the region.
- @Manishearth There is a parallel of there being 2 types of API users as there are 2 types of MessageFormat users: people who want to format individual messages vs people who want to do bundles and load many at once
- @robertbastian I don't think that is analogous at all.
- @sffc That isn't the first analogy I would draw. The question of display names applies to currencies and units as well, or any case with a lot of identifiers. An approach that we have partially implemented for this is to have a wway to query the data marker attributes, by doing prefix matching. We had to do this for loading the model for Segmenter. We can iterate over all available regions at runtime.
- @robertbastian I don't think we ever iterate. Iterate would require a b-tree map, and that takes a lot of ___.
- @sffc I think Buffer will support it.
- @robertbastian I don't like the idea of loading N payloads. To do that, you need to allocate `DataPayload<>` into a `Vec`.
- @sffc Do we actually need that? Another design is, if you want a multi-{unit, region}, the formatter just saves a reference to the provider.
- @robertbastian Then it would have a lifetime, which would be a first for us to do so. The last time when we designed a display name formatter for a single region, and yesterday we said that we can do that. In DateTime, you can't create a DateTimeFormatter per time zone.
- @sffc I think this is a policy decision.
- @robertbastian It's one of the 3 options on the table. The other options are heap allocations or formatters with lifetimes on the providers.
- @sffc I think those options are in the same bucket. I want to evaluate the technical merits of the options first, and then design the API after that.
- @sffc One piece of data that we don't have is the degree to whicih the data size difference between data markers & ___ .
- @robertbastian It depends on the user's calling code. If the caller only ever formats `GB`, one wins over the other. But if all options are equal, then it doesn't make a difference.
- @sffc There's loaded data size and linked data size. What is the impact on the linked data size? The linked data size still depends on auxiliary keys. None of the proposals on the table use actual data markers on the region. You have to do slicing, but you can do slicing manually.
- @robertbastian With finely-sliced data markers, you can do filtering on the data keys. But with multiple payloads, you can't do such filtering.
- @Manishearth We can revisit the policy about auxiliary keys. Part of the issue is that finely-sliced data doesn't scale to compile. It takes gigabytes of RAM to work, and we see this on CI. It seems to be a function of the number of tokens.
- @Manishearth I don't think we're aligned on what the actual problems are constrains are.
- @sffc ZeroTrie indexes into a static array. If it were to index into a ZeroVec, that would solve the problem. It's one of the points that I listed in issue #2230.
- @sffc I want to discuss the merits of the two approached (finely-sliced data markers & load multiple payloads)
- @Manishearth Yes. We're talking about the relative importance of these use cases. Between "format a specific set of locales" vs. "format any and all locales", do we have a preference? is there a tradeoff?
- @sffc Two things to add, not completely related: For display names and probably for units, the model that finely-sliced would push us towards where the unit / region would be in the constructor would be okay. In the case of units, it's generally the case that the measure unit in the constructor. In the case of display names, in general, you will want multiple of them. But display names are so thin, so just use your data provider and provide it to your static function (? constructor?).
- @robertbastian You can skip the formatter thing and go straight to the formatted thing.
- @sffc My second statement, unrelated to the previous, is that I think we still have headway to make multi-payload to be competitive with the link size we can get finely-sliced. When I say link size, I assume that you're not doing custom slicing. Assuming that you link all regions and all locales, I think there's headway to make the multi-payload competitive. Every locale has a base and overrides. Overrides are where the locale differs from the base. That allows locales to share the Latin names and only store the difference when they differ. This helps low resource languages that don't have complete data yet, so they default to the root locale that has the Latin names in those cases of missing data. We have an exception to ___ when it's a performance related thing. If we ship the Azeri pack but only included 5 region names and not the other 30, then it would fall back to root. It somewhat harms the i18n, but not to the degree that we couldn't make a policy that this is a cost that we're willing to accept.
- @Manishearth Every locale has a base and override. Is the base the root locale? Or does every locale have it's own base?
- @sffc The base for a locale is the script-based.
- @robertbastian I think what you're designing with the multi-payload solution is a workaround. I think finely-sliced data markers would allow us to get full deduplication. So if we can do that, we should.
- @Manishearth I don't think it's entirely a workaround. We never designed our data loading infrastructure to have more than one payload. It is theoretically possible to design a data provider to load two separate keys referring to differing size subsets of the same data.
- @sffc Showing data from `provider/source/data/debug/datetime/LocationsRootV1` and `provider/source/data/debug/datetime/LocationsV1`. Example: the `fr` locale has ~400 lines in `LocationsRootV1` but only 223 lines in `LocationsV1`. A lot of the locales in `LocationsRootV1` have identical contents, so they get deduplicated, and we only store one copy.
- @Manishearth Can we design finely-sliced data markers so that they support multi-payload loading?
- @robertbastian Instead of passing around a CountryFormatter, you'd pass around your data provider and do on-demand loading. You wouldn't store a `Vec` of `DataPayload`s, but you need to build some kind of map, so effectively you're rebuilding the data provider
- @Manishearth We've previously established that data loading is supposed to be cheap, so that seems fine.
- @sffc The only downside that I can think of is the iteration over all of the things in the provider is not something that we support.
- @robertbastian We have that, it's just behind the `alloc` feature. We can make it no-alloc
- @Manishearth We've cared about the `alloc` stuff for a while.
- @sffc One of the reasons that we haven't had the iteration provider because we don't want users to use that in order to determine what locales are supported.
- @Manishearth Do we need iteration over locales for display names?
- @sffc Having a region picker is the main use case.
- @robertbastian But is iteration over display names the right way to do that? We dont do that for time zones, we iterate through time zone ids, not time zone display names
- @robertbastian Not all locales have display names for all regions. So you can't simply iterate over display names.
- @Manishearth Yeah, so the list of regions could be either "the regions for my app" or "all regions ICU4X supports".
- @robertbastian Something like "known regions".
- @sffc In general, ICU4X compiles with all regions, and you can slice which regions you want.
- @robertbastian All regions changes with every CLDR release.
- @Manishearth Fallback also affects which locales we choose. These are interesting questions, but they don't tie into display names, they tie into a different issue. We should file an issue for this and talk about it separately.
- @sffc ___ has an API `supportedValuesOf`. Maybe the basis is the LocaleCanonicalizer.
- @robertbastian I don't think LocaleCanonicalizer is the struct to do this.
- @sffc Or maybe the struct is LocalExpander. The higher level point is that this example can be a basis for display names.
- @manishearth Also while I'm not a fan of doing ECMA specific APIs we could have a DisplayNamesSupportedValues marker (etc). It's locale unaware anyway, so doing this API based on a locale is incorrect: we can have a big vec of stuff.
- @sffc One other problem with finely-sliced, we generally index each one of them on the region code, which is two letters. But then we have to repeat the name of the unit in the name of the key along with the locale when we load units. So we end up with a key like `en-kilometer`
- @robertbastian Display names change when CLDR changes display names. Units change when ISO changes them. So they shouldn't be tied together. You should not checksum them together and use one to look up the other. We don't do this in time zones. The display names are not checksummed to the IANA mapper.
- @sffc I am trying to parse @robertbastian's comment because I think checksumming is only feasible in the multi-payload approach, and I think I described that. I don't know why you say it's not feasible.
- @robertbastian We didn't want to load the IANA mapper.
- @Manishearth I think this is possible with the finely-sliced approach. I think it would work as such: If you have the locale French, then you would get the display names map without auxiliary keys. It would use the index numbers as the keys, where there is a side map that correlates the index number to the actual key name. This needs the checksum to be 100% correct. It's not my preferred, but we can do that. We can use at most 2 bytes for the index number.
- @sffc It may not benefit display names, but units would benefit from using 2 bytes.
- @Manishearth Units get added over time.
- @sffc I like the checksum approach.
- @Manishearth Turning this into numbers is okay.
- @sffc This sounds like a code driven thing, and I would like things to be data driven.
- @Manishearth This is a small amount of data.
- @Manishearth This would give our `supportedValuesOf` API for free.
- @robertbastian I'm not a big fan of checksumming markers together. If we can put this into a special compression locale for teh same marker, it would just need to be in teh same data struct somehow.
- @Manishearth For datagen specifically, we can tweak it, instead of doing the regular databake implementation, call this special one for this one key. It produces two ZeroTries, one with the metadata, and one with the special stuff for the key. Doing it in datagen would work, but it would not be so easy in Postcard.
- @sffc Another approach without doing the checksum, as long as BakedProvider or Postcard know the full space of all possible attributes, is that they can do this table thing internally and deal with during the load operation. Not sure if that was what @Manishearth was saying or not.

```rust
// singleton, checksummed
struct Compression {
   tzids: ZeroVec<TimeZoneId>, // [deber, usden]
}

// compression locale
struct DisplayNames {
    names: ZeroMap<TimeZoneId, str> [(deber, ""), (usden, "")]
}
```

- @robertbastian At some point you can just run deflate on the payload bytes instead of inventing our own compression schemes
- @Manishearth It sounds like we mostly agree on finely sliced? There's just an inefficiency we want to handle, we know we can handle it with a split metadata table, and the disagreement is between weht that table to be a singleton or purely internal in baked and postcard.
- @sffc I'm onboard with finely sliced on the condition that we fix #5230.
- @robertbastian Only thing we lose is the ability to filter regions on datagen time.
- @sffc We can do it in multi-payload. There would be a three-layer approach. A singleton `und` locale would contain a display name for every region. And then the `Latn` and `fr` locales could choose to strip some of the display names. The observable behavior is that all of the display names are supported, but the i18n is less good, which we could decide is consistent with our policy.
- @robertbastian Why are we discussing this compression a lot? ZeroTrie is efficient, yes?
- @sffc Manish asked what the challenges were: I pointed out that the multipayload approach allows for the attributes to be better compressed via the checksum approach, but now we have a path for that compression in both models.
- @manishearth: Current impression is that everyone is on board with finely sliced, provided we fix #5230. We have a path towards compression with two routes (explicit metadata singleton vs internal stuff), with pros and cons that can be discussed. We have ideas for #5230.

Discussion on #5230:

- @Manishearth If the data is a simple string, baked data can optimize how it's stored. The official data struct could be a repr(transparent).
- @robertbastian Now we can make anything the data struct.
- @sffc DisplayNames, maybe, but I want this to work in general for finely-sliced data including currencies which have a more complex data payload.
- @robertbastian The optimization could work for `Cow<'static>` something. Which can include both strings and patterns.
- @sffc This requires us to use the Yoke code path in DataPayload. Which might make "alloc" features harder to add.
- @Manishearth We've resolved that problem. But we need to use ZeroFrom. It's unfortunate that we can't use StaticRef; but I'm okay with that.
- @Manishearth I think there's one cost left. The Yoke code path may still link in a Drop impl, which could include panics and things. Code that used `from_static_ref` linked a constructor.
- @robertbastian We could add a third variant to DataPayload. The Yoke variant is being abused as the Owned variant. This can be Copy if we have a Cow where we can turn off the alloc variant.
- @Manishearth The specific cost is, even in a crate that is using `alloc` but cares about code size, Drop and panic impls get linked. Which can be a code size bloat that could be significant. One of the answers is, don't use alloc if you care about code size. I don't think we can solve the problem by adding a third variant to DataPayload. The destructor of Rc is also a problem, but it can be improved a bit. That's something that can be fixed upstream.
- @sffc Another approach is for finely sliced data markers to specify the owned and borrowed versions separately.
- @robertbastian We don't bake the owned, we stuff it into a zerovec and bake the zerovec.
- @sffc We end up with a static reference to the ULE type not the data struct value type
- @robertbastian: rn icu_provider is independent of ULE shenanigans
- @sffc: DataMarker gains a new optional field for the borrowed variant. providers that benefit from that can use it 
- @manishearth not a fan of adding a new variant
- @sffc solution has some merits, cost is it makes our type system a bit more complicated.
- @manishearth i think we have a solution here already, the downside of the solution is "yoke cost" which we've solved in no_alloc and is _mostly_ gone in alloc except for causing a small codesize bloat in the situation where you have enabled debug assertions in std (unusual), care heavily about codesize but have alloc enabled (unusual) and ar eavoiding linking in panic formatting code.
- @sffc We know that `ZeroFrom`, `DataPayload::get`, and `Cow::get` are not completely free. We could, in principle, avoid all of those costs if we added a `Borrowed` associated type.
- @Manishearth I think these costs are very minor relative to a DataProvider load operation.
- @sffc The DataPayload load operation can be quite cheap, especially if we give it certain optimizations.
- @Manishearth But still on a different order of magnitude than these pointer indirections.
- @robertbastian The reason you may think get is expensive is not because of the discriminant/branch, it's expensive because the owned data payload cannot perform inlining across the get, which you can do with static. That's why we have borrowed. This is solved by borrowed variants, doesn't need to be solved in data payload any more.
- @manishearth: Solution: we could make bake export a function on data markers that have this optimization applied.
    - Example: `pub fn load_direct_display_names_v1(req: DataRequest) -> &'static str`
- @sffc How do we get the type `&'static str`?
- @Manishearth It can be a macro thing.
- @manishearth in DislayNames we dn't even need to do DisplayNamesBorrowed, we can just have this be a pub function.

Plan for #5230:

- Have a "varzerovec optimization" for databake that instead of doing the regular array of data structs, it does a single varzerovec of data structs. The load impl converts to a yoke, which is not much of a cost, and not a cost at all in no_alloc
- This optimization is triggered on a hardcoded list of concrete data structs (Cow<'static, str>, ZeroCow<'static, Foo>, ...) in the baked exporter, using Any downcasting to obtain the ULE bytes.
- This optimization never triggers on "empty" (no locales/data) datagen invocations, like stubdata.
- Future improvemnet:
    - The optimization can also generate a `Baked::load_direct_{data_marker}()` function that can be used for borrowed-variants or for direct-loading free functions. This is not necessary at the start but can be added as necessary.
    - For handling the "empty" case we may need to hardcode markers, or have this in DataMarkerInfo. We're not fully agreed on the best way to do this.
    - Note: we do have code currently assuming the variant of the datapayload and unwrapping, this would heuristically change the variant that is returned from `load`. We will need to be aware of this if we go ahead with this optimization

Agree: @manishearth @robertbastian @sffc

Possible plans for key table optimization (#3260):

We have two optimizations possible. Both optimizations involve using a multi-level optimization where the actual aux keys stored in the trie are integers instead of region codes or units, and we use a separate shared singleton lookup table that maps aux keys to integers that get stored in the trie

1. Do not do anything special: let ZeroTrie solve it
2. Apply the side-table optimization internally in databake and postcard.
    - It may need BlobSchemaV4. Downside: new schema.
    - Or we could store the side-table in a `\0t` key, for example.
3. Use a checksum, with a singleton key defining the key table. The logic will have to be done manually in the component. Downside: opaque to filtering

**Comment by Manishearth** [comment]
Triage: In 2.2 as long as DisplayNames is a headline item for 2.2

**Comment by sffc** [comment]
Further discussion of the design of display-names-flavored data provider architecture:

A key use case to support is one where you need to load multiple display names in the same locale, such as building a drop-down menu for selecting your region, time zone, etc.

Clients can load each marker/locale/attribute separately with individual load operations. But, in most of our providers, this is costlier than loading the marker/locale first and then loading the attributes as a second operation, re-using state from the initial marker/locale data load.

Here are four architectures to achieve this optimization.

1. Intermediate type generic on data provider, normal provider internal layout
2. Intermediate type with no generic, normal provider internal layout
3. Intermediate type with no generic, required marker-specific internal provider layout
4. No intermediate type

## Architecture Details

### Architecture 1

In #7621, I've implemented this architecture.

The types would have roughly this shape:

1. Compiled data: `MultiDisplayNames::try_new(locale) -> MultiDisplayNames<BakedBound<M>>`
2. Blob data: `MultiDisplayNames::try_new_with_buffer_provider(provider, locale) -> MultiDisplayNames<BlobBound<M>>`

The data provider bound in `try_new_with_buffer_provider` (and `try_new_unstable`) is not the normal data provider trait; it is a new trait that returns the intermediate provider type instead of a payload.

On `MultiDisplayNames`, you can call `.load("US")` to load the display name for that territory, and it will use a code path statically optimized for the data provider flavor.

Pros:

- Builds on top of all previous decisions about data provider layout, including those earlier in this thread
- Data providers that cannot implement the trait efficiently, perhaps FsDataProvider, do not need to implement the trait. They just won't support this optimization. Clients can still load DataIdentifier by DataIdentifier.
- Gives leeway to individual data providers to make their own decisions.
- Works nicely with forking providers.
- Requires no changes to datagen.

Cons:

- The intermediate type has a generic, but we don't like generics
- The intermediate type retains a chunk of the provider, but we like the provider to be transient
    - NOTE: It need not retain a _reference_ to the provider, since it can yoke itself to the backing buffer, same as DataPayload.
- Over FFI, requires either multiple types or a single type with a (feature-gated) branch.
- Multiple new traits and things exported in `icu_provider`

### Architecture 2

This is related to architecture 1, except the intermediate type does not have a generic.

1. Compiled data: `MultiDisplayNames::try_new(locale) -> MultiDisplayNames`
2. Blob data: `MultiDisplayNames::try_new_with_buffer_provider(provider, locale) -> MultiDisplayNames`

In order for this to work, we need a trait where providers convert their internal layout into a single public canonical layout. That layout might be something like `(ZeroTrie, Cow<[M::DataStruct]>)`, for example. We would choose a layout that is optimized for compiled data and might require allocation and eager deserialization for blob data. Alternatively, we could choose a layout that is an enumeration over all of our stock providers, with feature-gated branches.

Pros:

- Builds on top of all previous decisions about data provider layout, including those earlier in this thread
- Data providers that cannot implement the trait efficiently do not need to implement the trait.
- Translates seamlessly to FFI.
- Requires no changes to datagen.

Cons:

- Pessimizes blob data, or makes compiled data slower.
- The intermediate type retains a chunk of the provider, but we like the provider to be transient
- Encodes a stable public layout for the intermediate type, making it more disruptive to tweak the layout if we optimize the data provider internals or the ZeroTrie
- Does not work well with a forking provider. If you have two providers, each with half of the attributes, you would need to merge them at runtime into the single canonical layout.
- Multiple new traits and things exported in `icu_provider`

### Architecture 3

This option, an evolution of a design conceived by @Manishearth, involves annotating a data marker with the expected memory layout of that data marker. Data providers must store the data with that layout, and the intermediate display names type can then use that one single memory layout.

For example, a data marker could get a new annotation `use_multi_attribute_layout: bool`. If the annotation is true, data providers must be able to return a `MultiDataPayload<M>` of a prescripted layout, probably a ZeroTrie of attributes and a Vec of values. With this information, data providers can choose to make different decisions on how they store the data. For example, FsDataProvider could store all the attributes together in a single file, and BlobDataProvider could store a VarZeroVec of parsed values instead of a slice of serde buffers.

This option still needs a runtime trait like `MultiDataProvider`.

Pros:

- Data providers can use the knowledge that a marker is multi-attribute to optimize its storage
    - NOTE: We could still add the flag in options 1 and 2 as a hint, without making the layout be a requirement.
- By way of semantics, we are not retaining a chunk of the provider; we are retaining a multi data payload based on a contract with the MultiDataProvider trait.
    - NOTE: Mechanically under the hood, this architecture is basically identical to options 1 and 2, at least for compiled data. This "pro" is describing the clarity of the contract. _This is a key point!_
- Data providers that cannot implement the trait efficiently do not need to implement the trait.
- Translates seamlessly to FFI.

Cons:

- Requires changes to the data exporter, which is already quite complicated, and probably to the blob schema
- Does not leverage the VarULE layout optimization previously agreed in this thread. However, it's possible that we could consider `use_multi_attribute_layout` to be a _replacement_ for the VarULE layout optimization.
- Encodes a stable public layout for the intermediate type, making it more disruptive to tweak the layout if we optimize the data provider internals or the ZeroTrie
- Does not work well with a forking provider. If you have two providers, each with half of the attributes, you would need to merge them at runtime into the single canonical layout.
- Turning on or turning off the `use_multi_attribute_layout` flag is probably a breaking change, since it changes the layout of the data in the provider, so we need to be confident ahead of time that we want to enable the flag for a particular marker
- Multiple new traits and things exported in `icu_provider`

### Architecture 4

Here, we do not have an intermediate type. We take the provider and a list of attributes, and we directly return a collection of resolved values. We could return a `Vec`, an opaque container type, etc.

I am not considering the variation where we return an external iterator, because the iterator is itself an intermediate type, and then we need to answer most or all of the same design questions as in architectures 1, 2, and 3.

To avoid the allocation, we could consider an internal iterator, where we take a `FnMut` and call it for each value in the collection. For example, the signature of the function could look something like,

```rust
DisplayNames::load_each(&["US", "CA", "MX"], (name) => { ... })
```

Mechanically, I would probably implement this similarly to how I implemented option 1, but I would keep more things internal.

Pros:

- Builds on top of all previous decisions about data provider layout, including those earlier in this thread
- Don't need to expose any internals to the user. We retain full flexibility.
- Allows providers to do even more advanced optimizations for loading a list of attributes.
    - NOTE: The intermediate types in architectures 1, 2, and 3 could have `load_into_vec` functions that also could allow for these types of optimizations.
- Gives leeway to individual data providers to make their own decisions.
- Easy to implement across arbitrary data providers, including forking providers.

Cons:

- Very narrow to this one use case. Architectures 1, 2, and 3 could be used in other situations, like unit and currency formatting, where the formatter may want to retain the multi-attribute structure as a field.
- Unless we want to require diplomat callbacks, FFO probably requires allocating and returning the collection.

---

My thoughts after having written this up:

- I don't see any good reason to pick option 2 over option 1, but I'm not allergic to funny generics.
- Only options 1 and 4 support forking providers. I feel like this is an important use case, and not supporting it is a bad decision. For example, it seems totally reasonable to have two data providers, one with Tier 1 territory names and one with Tier 2 territory names.
- We could start by implementing option 4, and expose the intermediate type later. If we can't agree on an architecture, I suggest that we start with option 4. However, if we can agree on 1, 2, or 3, we could add option 4 as a convenience function later.

**Comment by Manishearth** [comment]
Option 4 is interesting, but I feel like it will end up doing a lot more allocation. I also think in most cases you just want the ability to query "all" and that's something we would have to design.

I think it's something worth exploring but it doesn't strike me as a good "default".


I don't consider traits to be a major cost in `icu_provider`. Overall I'd like to avoid generics in persisted types where possible, but also this is ultimately a power user API so maybe it's fine.


> I am not considering the variation where we return an external iterator, because the iterator is itself an intermediate type, and then we need to answer most or all of the same design questions as in architectures 1, 2, and 3.



Though we can avoid persisting the iterator, so an intermediate associated type works.

**Comment by Manishearth** [comment]
One question: what does the Baked version return? Because `MultiDisplayNames<Baked>` isn't a type that can be named stably.

Is there a use case for a multi-type with Baked? It does save on lookups, I think.

----------


We should probably write out the use cases for these types. I think all of these designs seem pretty flexible in how the final formatter/etc type can choose to use them.

For display names it's mostly for formatting a dropdown, where you want `n` display names. A vector based model isn't terrible there.


On the other hand for currencies people may be using something lazier, maybe there isn't even that much of a use case for multi providers since you have to still format individual numbers (unlike DisplayNames, which are just strings we can shove in an array).

Unit formatting _seems_ similar to me, but I'm not sure.

**Comment by sffc** [comment]
> One question: what does the Baked version return? Because `MultiDisplayNames<Baked>` isn't a type that can be named stably.

It can just use a public `CompiledData` marker type exported in the same module as `MultiDisplayNames`, or maybe in a `scaffold` module.

> We should probably write out the use cases for these types. I think all of these designs seem pretty flexible in how the final formatter/etc type can choose to use them.

I've been predominantly focused on the Display Names use case. For currencies and units, there are some cases where you are displaying multiple different currencies or units on the same page, but mostly you use just one at a time, so the use cases are weaker. On the other side, we decided for time zones that you frequently need to display multiple time zones on the same page, so we didn't even try to support the "only one time zone" data loading model.

**Comment by sffc** [comment]
WG notes:

- @sffc introduces topic
- @robertbastian This is adding a lot of complexity for one particular use case (formatting 100 territory names at once) and it's not even clear that this gives a noticeable performance impact (you save like 5 steps in the ZeroTrie) – have you benchmarked that this is worth it? 
- @sffc don't have numbers
- @hsivonen from the firefox perspective: we bake in everything, but DisplayNames is an exception. Eemeli and I want to maybe do it more dynamic. The reason why it involves DNs being bad i18n: for territory names, the territory name list is politically sensitive. For the application it may be super important to be able to know what strings are going to come out. Makes territory names a bad idea to bake in
- @hsivonen For languages, we have the linguistic base form of each language name, but if you think of the context of a language name, you don't necessarily just say English, you may have a designer say "in English", and in a language other than English it will cause morphological change and then you can't use DNs anymore
- @hsivonen So who is this for? If we want APIs for good i18n ... browsers have this,and if people want to use dynamism instead of baked data... but from the POV of designing APIs or cases that are not tied by legacy, what is the new greenfield use case?
- @hsivonen I can see apps with niche use cases needing to render language tags into user readable form.\
- @sffc Ok so DisplayNames is a new ECMA API, after introducing addl requirements for motivation. TG2 fairly recently (last 5y) felt DNs will well motivated since apps were shipping large payloads for language pickers/dropdowns.
- @hsivonen apps want to choose from a wide variety of languages, as opposed to the 3/4 languages relevant to that domain.
- @manishearth clarity on 2.2 blocking?
- @manishearth I think this is decently motivated, this has cropped up for units, DNs, timezones, currencies ... But the use cases are different
- @sffc The design for returning a multi display names is the "normal thing" that we have always supported, ECMA has this, ICU4C has this. The "small" string API is the novel thing.
- @sffc You can think of this as an optimziation, an ECMA thing, etc. Lots of motivations here. Can look at alternatives, like string-returning APIs
- @sffc this is DisplayNames blocking, not 2.2 blocking
- @manishearth (getting clarity that this is the primary type for the crate)
- @manishearth if it is the primary type I am less enthused about the generic. But maybe it's fine.
- @sffc also want to point out I want to use this for currencies/etc
- @sffc note that if we had this we might have implemented timezones differently
- @sffc the thing this supports that tz doesn't is slicing. slicing out tzs is something some clients may want to do
- @hsivonen Noting that MSFT doesn't ship DNs for old currencies and it's not a problem
- @hsivonen for other categories there's a less clear line. do you have thoughts?
- @sffc thing with static slicing is it can only be coarse. have 2-3 buckets, which CLDR has defined. "display names for territories where the language is spoken in the territory", "all territories in modern use", "all territories incl those not in common use"
- @hsivonen now super curious about not-modern-use territories in CLDR
- @hsivonen feedback from me is I'm not currently able to eval these options, but unlike everything else where we use baked data, this is an area where we may want to do something more dynamic.
- @manishearth That's useful for design: it means that we have someone who wants somethign that supports bfufer well (maybe to the point of using Fork) and can be wrapped in ECMA.
- @Manishearth For next steps, shoudl we enumerate the use patterns? There's ECMA, there's dropdowns. For currencies, time zones, units, are all different. The fact that they're different will let us make progress.
- @robertbastian also get some numbers.
- @sffc tbh I'm not 100% sold on attributes, only 80% sold. If our choice here is that the infra is too complex then maybe coarse slicing is fine.
- @robertbastian: You don't need loading shenanigans to use marker attributes
- @robertbastian this is the original reason we added marker attrs, so you can do filtering in datagen
- @robertbastian we then used them in datetimes
- @robertbastian original design was you had a single call and it loads everything
- @manishearth (requesting clarity)
- @robertbastian wants to generate per-attribute markers but then either load one at a time or do a 100 loads ina row
- @sffc looks like Robert wants us to stop at the original attribute design, whereas to me it's an open question we decided to fill in later
- @sffc I don't see this as a performance optimization.

**Comment by sffc** [comment]
Another implementation of architecture 4 would be something I had mentioned a long time ago, even before we had attributes: use the data response metadata to save provider-specific indices into its zerotrie (or any other structure), which get passed down again through the data request metadata to jump directly to the attributes trie.

This implementation approach has advantages and disadvantages. I feel the disadvantages are stronger, but others might feel otherwise:

- Pro: No new traits; leverages existing machinery
- Con: Arbitrary garbage can be passed into the request metadata; creates more GIGO paths
- Con: Only makes sense with architecture 4, where a function reads all the strings from the DataProvider without retaining the provider
- Con: Seems tricky to implement on a forking provider

**Comment by sffc** [comment]
Design question:

We have a few dimensions to these display names. It's not completely clear which ones should land as markers, as attributes, or as values.

**Definitions:**

- Marker: for example, LocaleNamesRegionLongV1 vs LocaleNamesRegionShortV1 for long vs short
- Attribute: for example, "FR/long" and "FR/short" as the data marker attributes
- Value: for example, "United Kingdom" and "UK" stored as separate payloads vs ["United Kingdom", "UK"] as a value stored in a single payload

**Dimensions:**

- Length: long and short. Maybe more in the future.
    - Example: "United Kingdom" is long, "UK" is short.
- Context: standard and menu. The menu form is further broken down to core and extension.
    - Example: "Traditional Chinese" is standard, "Chinese, Traditional" is menu. (They do it this way so that it sorts next to "Chinese, Simplified" in a language picker.)
    - In the menu variant, "Chinese" is considered the "core", and "Traditional" is considered the "extension". CLDR stores these as separate strings that can be glued together.
- Coverage: common and comprehensive, or possibly three buckets, minimal, common, and comprehensive.
    - Example: languages that are spoken in a region have that region in their "minimal" set; all modern regions are "common"; and regions that are not in modern use can be "comprehensive".

There are some tests we can use to determine where the things should go:

1. Is it a developer preference or a user preference?
2. How important is it to be easy to slice?

Answers to these questions:

- Length: Developer preference. Somewhat valuable to slice, except there is not a lot of short data. Most regions are long only. Only a handful have short forms.
- Context: Developer preference. Similar to length, there is not a lot of menu data. Most languages are the same in their standard and menu variants.
- Coverage: Mostly a developer preference. This is the one with the _most_ slicing potential.

This would suggest that coverage is the most important thing to put in markers. Length and Context could go in markers, or we could stuff them in attributes or values. If we stuff them in attributes, they are still sliceable at datagen time, but they would want a regex for slicing since the attributes also contain the language / script / region code.

Thoughts?

**Comment by Manishearth** [comment]
> Marker: for example, LocaleNamesRegionLongV1 vs LocaleNamesRegionShortV1 for long vs short


A thing I noted in discussion today was that we have a series of options here. We can have three markers (common, long, short), where everyone benefits from slicing. Or we can have two with fallback (long, short falls back to long; OR short, long falls back to short). Of those two options the former seems to make sense. Or we have one combined marker.

Given your comments about length, "not a lot of short data" seems to indicate that we should go with one combined marker.

**Comment by Manishearth** [comment]
> This would suggest that coverage is the most important thing to put in markers. Length and Context could go in markers, or we could stuff them in attributes or values. If we stuff them in attributes, they are still sliceable at datagen time, but they would want a regex for slicing since the attributes also contain the language / script / region code.

I think this is a good answer. I think context should also be markers since there's only one key that cares (unless I'm wrong?).

I would kind of prefer to avoid "nested" attributes but that's a very weak uninformed preference, and it's mostly from not being sure about how well that performs with various zero trie optimizations.

**Comment by sffc** [comment]
We discussed this in the WG meeting with @Manishearth @robertbastian and myself. My summary:

- Coverage is something we all agree should be a marker-level distinction.
- Unclear if we need 2 or 3 levels for coverage.
- Context likely needs a different data struct, since CLDR ships the data as menu=core and menu=extension. This forces our hand into making it be a marker-level distinction.
- Length is a bit of a gray area. We looked at the current data sizes, and both long and short add only 1% each to the overall display names data size. We are at this point in time leaning toward making them attributes.

**Comment by sffc** [comment]
Deliverables for 2.2:

- [x] Land #7697 and implement for all other data markers we currently have
- [x] Keep the existing old "multi" display names code; organize it and the "uni" display names code into separate modules, with documentation on why clients should pick one or the other for beta testing purposes, with a link to a new issue seeking feedback
- [x] Document the limitation on locale display names with a link to an issue for implementing the full algorithm

**Comment by sffc** [comment]
Notes from a conversation with me and @Manishearth 

> @sffc - The old data put "en" in a marker called Language and "en-GB" in a marker called Locale. This was useful when we used Language subtag as a key in the ZeroMap, but that is not a consideration in the attributes design. Do you think this is a useful distinction for users?
> I'm going to proceed with the assumption that this is NOT a useful distinction in the new model. Ok, then we have the follow-up question, is a LanguageDisplayName a useful type, or should it be just LocaleDisplayName? Should LanguageDisplayName work perhaps only on a LanguageIdentifier?
> ECMA-402 type "language" handles a LanguageIdentifier. That seems like good enough precedent for how our LanguageDisplayName type should behave.
>
> @Manishearth it does not seem to be a useful distinction to me. i feel like the point is still what people call languages. but idk. i think it should probably only work with LanguageIdentifier

Given this, I don't have time to finish LanguageDisplayName for 2.2, since it requires all the extra work with parentheses and such. We have RegionDisplayName and ScriptDisplayName which should be enough to get early feedback, acknowledging that language display names are the most "popular" type, which we aren't supporting yet.

**Comment by Manishearth** [comment]
Sounds good! Would you mind adding the right wording for this to the release notes? https://docs.google.com/document/d/1U-obZvYZD56addIFBDPuhJwYLwAxQfWDobw-_KbF_aU/edit?tab=t.0

And in that case are there any remaining items for 2.2 for displaynames, or can we consider it done?

**Comment by sffc** [comment]
no this is the first bullet point. The other two are moving things around and adding docs. I'm working on that pr... #7826

**Comment by sffc** [comment]
I think #7826 resolves the bullets from https://github.com/unicode-org/icu4x/issues/3260#issuecomment-4122155905. Someone (probably me, could be someone else) needs to scrub this issue for any remaining open items and spawn follow-ups, then we can close this issue.

**Comment by sffc** [comment]
New issues spawned:

- https://github.com/unicode-org/icu4x/issues/7831
- https://github.com/unicode-org/icu4x/issues/7825
- https://github.com/unicode-org/icu4x/issues/7833
- https://github.com/unicode-org/icu4x/issues/7824
- https://github.com/unicode-org/icu4x/issues/7832

One action left for 2.2 here: fix the docs that landed in https://github.com/unicode-org/icu4x/pull/7826. They say that `multi` has limited support but it is `single` that has the limited support. The issue https://github.com/unicode-org/icu4x/issues/7825 has the correct info.

**Comment by Manishearth** [comment]
I believe this issue is resolved as far as 2.2 is concerned .
